lib/Sema/CSSimplify.cpp

```diff
@@ -7750,7 +7750,20 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     if (type1->isPlaceholder() || type2->isPlaceholder())
       return getTypeMatchSuccess();
 
+    // If the other side is a tuple containing a pack expansion with the same
+    // number of elements, record a fix.
     auto *loc = getConstraintLocator(locator);
+    if (loc->isLastElement<LocatorPathElt::ApplyArgToParam>()) {
+      if (auto paramPack = simplifyType(type2, flags)->getAs<PackType>()) {
```

Comment:
Do we need to call `simplifyType` here? At the beginning of this method `getFixedTypeRecursive` is called which is going to expand pack expansion types wrapped in tuples.

Reply:
No, it seems like `type2` is a pack expansion type, so we need to simplify it to get the number of elements in the parameter pack. Tests are coming!

Reply:
Okay may how about:

- if `type2` is a PackExpansionType take `getCountType()` from it, check whether it's fully resolved (`!hasTypeVariable())` and use that to get element count if it's a PackType (note that pack expansion could also be for pack archetype which means that its could is going to be PackArchetypeType).

- `type1->getAs<TupleType>()->getNumElements()` needs to be split into:
```
if (auto *tuple = type1->getAs<TupleType>()) {
   if (tuple->getNumElements() == ...) {
      ...
   }
}
```

because if `type1` is not a TupleType `getAs` would produce `Type()` and `getNumElements()` is going to crash.

Reply:
> check whether it's fully resolved (!hasTypeVariable())
`hasTypeVariable()` returns `true`, so this check does not work.
> pack expansion could also be for pack archetype
If it's a pack archetype, it does not have a way to check the number of elements in it.
Also, we cannot avoid simplifying the type. Even with count type, I had to simplify it because it is `type_variable_type` that requires simplifying to get the number of elements

I added your other suggestions, though!

Reply:
There is a difference between trying to simplify whole pack expansion type and only its count type, I think simplification of count type is okay but what I was trying to hint at is a bigger problem - even though we might have a pack type for shape of the expansion there is no guarantee that it's a fully resolved one at this point and if it's not, there is no way to determine the actual number of elements in the pack expansion when it contains other pack expansions as elements. Same goes for the tuple type. 

If we consider that the diagnostic is about the expected number of arguments I'd say that we should relax the equality check here and turn it into `if (type1->is<TupleType>() && paramPack.getNumElements() > 1)` this would cover the case where some of the elements are not yet resolved and would produce a reasonable initial diagnostic to hint users to drop parenthesis and subsequent errors would diagnose any and all type and arity mismatches. Would be great to add a couple of tests for that too.

Reply:
> If it's a pack archetype, it does not have a way to check the number of elements in it.

Right, I was just pointing out that either is possible, we cannot act in cases where shape is an archetype because we don't know the number of elements statically.

Reply:
Is this a bug? 
```
(lldb) po simplifyType(packExpansion)->dump()
(pack_expansion_type
  (pattern=pack_archetype_type address=0x126815ad8
    (interface_type=generic_type_param_type depth=0 index=0 decl=test.(file).foo(u:).U@/Users/simanerush/Apple/Swift-Compiler/Swift-Files/test.swift:22:15 pack) name=each U
)
  (count=pack_archetype_type address=0x126815ad8
    (interface_type=generic_type_param_type depth=0 index=0 decl=test.(file).foo(u:).U@/Users/simanerush/Apple/Swift-Compiler/Swift-Files/test.swift:22:15 pack) name=each U
))

(lldb) po simplifyType(packExpansion->getCountType())->dump()
(pack_type num_elements=1
    (pack_expansion_type
      (pattern=pack_archetype_type address=0x126815ad8
        (interface_type=generic_type_param_type depth=0 index=0 decl=test.(file).foo(u:).U@/Users/simanerush/Apple/Swift-Compiler/Swift-Files/test.swift:22:15 pack) name=each U
)
      (count=pack_archetype_type address=0x126815ad8
        (interface_type=generic_type_param_type depth=0 index=0 decl=test.(file).foo(u:).U@/Users/simanerush/Apple/Swift-Compiler/Swift-Files/test.swift:22:15 pack) name=each U
)))
```

Reply:
Hm, this might be a bug in TypeSimplifier (located in ConstraintSystem.cpp)?  I don't know how we end up with a pack expansion while simplifying a pack archetype...